--- uboot/oem/ami/fmh/hrotcore.c	2021-11-26 19:49:27.174868979 +0530
+++ uboot_new/oem/ami/fmh/hrotcore.c	2021-11-26 19:51:15.147135202 +0530
@@ -1,9 +1,14 @@
 # include <common.h>
+# include <stdlib.h>
 # include <config.h>
 # include <command.h>
 # include <flash.h>
 # include "cmd.h"
 # include <oem/ami/fmh/fmh.h>
+
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT
+    #include <environment.h>
+#endif
 # include "hrot_ifc.h"
 #include "u-boot/sha256.h"
 #include "u-boot/rsa.h"
@@ -23,6 +28,9 @@
 static sha256_context ctx;
 static UINT32 boot_fmh_location,product_info_offset=0;
 static char placeToHash[HASH_BLK_SIZE];
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT
+    static UINT32 totalsize=0;
+#endif
 
 void
 hrot_start(){
@@ -96,7 +104,10 @@
                         sha256_update(&ctx, hash_data, hs_size);
                         hash_start += hs_size;
                 }
-
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT		
+		totalsize=totalsize+hs_size;
+		env_set_hex("blob_length",totalsize);
+#endif		
 	}
         return 0;
 }
@@ -124,6 +135,11 @@
         char    linebuf[16384] = {0}, linebuf1[1024] = {0};
         void *buf=NULL, *buf1=NULL;
         int ret =0;
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT	
+	char blob_data[65]= {0};
+	int i=0, retval=0;
+#endif	
+
 		int sig_length;
 		char identifier[8];
 		struct spi_flash *flash;
@@ -150,6 +166,9 @@
 			printf("unable to find signature start identifier :( \n");
 			return -1;
 		}
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT
+		env_set_hex("blob_base",sigaddr);
+#endif		
 	
 		sigaddr +=sizeof(START_IDENTIFIER);
 		/* scan through hash area and find the proper hash for firmware image*/
@@ -194,6 +213,20 @@
                 printf("%s", key_data);
                 printf("%s", sig);
         }
+
+#ifdef CONFIG_SPX_FEATURE_BMC_DATA_FOR_TPM_SUPPORT	
+        for(i=0;i<size/2;i++)
+        {
+                retval=snprintf(blob_data+(i*2),3,"%02x",output[i]);
+                if(retval < 0 || retval >= 3)
+                {
+                        return retval;
+                }
+
+        }
+        env_set("blob_data",blob_data);
+        env_save();
+#endif	
         ret = rsa_get_pub_key_uboot_verify(key_data, sig, sig_length, output, size);
         return ret;
 }
