diff -Naur uboot_old/Makefile uboot/Makefile
--- uboot_old/Makefile	2023-01-11 12:22:06.299451983 +0530
+++ uboot/Makefile	2023-01-11 12:23:07.988077635 +0530
@@ -378,6 +378,7 @@
 
 KBUILD_CFLAGS   := -Wall -Wstrict-prototypes \
 		   -Wno-format-security \
+		   -Wno-address-of-packed-member \
 		   -fno-builtin -ffreestanding $(CSTD_FLAG)
 KBUILD_CFLAGS	+= -fshort-wchar -fno-strict-aliasing
 KBUILD_AFLAGS   := -D__ASSEMBLY__
