--- linux-5.4.99/drivers/net/ethernet/pilot_tulip/Makefile	2021-03-23 14:46:24.316839195 +0800
+++ linux/drivers/net/ethernet/pilot_tulip/Makefile	2021-03-19 17:18:45.175950000 +0800
@@ -0,0 +1,13 @@
+#
+# Makefile for the Linux "Tulip" family network device drivers.
+#
+EXTRA_CFLAGS += -Dfinal_version
+
+obj-$(CONFIG_PILOT4_NET_TULIP)		+= tulip.o
+
+# Declare multi-part drivers.
+
+tulip-objs			:= eeprom.o interrupt.o media.o \
+				   timer.o tulip_core.o		\
+				   21142.o pnic.o pnic2.o pilot_phy.o
+
